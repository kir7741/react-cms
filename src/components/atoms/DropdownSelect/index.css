.dropdown-select-field {
	display: grid;
	grid-auto-flow: column;
	position: relative;
	border: 1px solid var(--primary-gray);
	padding: 10px;
	border-radius: 6px;
	background-color: #fff;

	& > .icon {
		display: grid;
		position: absolute;
		top: 50%;
		right: 10px;
		transform: translateY(-50%);

		& svg {
			transform-origin: center center;
			transition: all 0.5s;
		}
	}

	& > .rotate svg {
		transform: rotate(180deg);
	}

	& > input {
		border: none;
		padding: 0;
		background-color: transparent;
	}
}

.dropdown-panel {
	position: absolute;
	top: 110%;
	right: 0;
	left: 0;
	margin: 0;
	padding: 0;
	border-radius: 6px;
	list-style: none;
	background-color: #fff;
	box-shadow: 0 2px 1px -1px rgb(0 0 0 / 20%), 0 1px 1px 0 rgb(0 0 0 / 14%), 0 1px 3px 0 rgb(0 0 0 / 12%);
	max-height: 160px;
	overflow-y: auto;
	z-index: 10;

	& > li {
		padding: 10px;
		min-height: 40px;
		cursor: pointer;

		&:hover {
			background-color: var(--ligth-gray);
		}
	}
}
