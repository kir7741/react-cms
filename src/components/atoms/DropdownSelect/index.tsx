import React, { useMemo } from 'react';
import classnames from 'classnames';
import ExpandDown from 'images/icon/expand-arrow-down.inline.svg';

import { OptionBase } from 'types/interfaces/option-base';

import Dropdown from '../Dropdown';
import styles from './index.css';

interface valueComponentProperty {

	/**
	 * 選中的文字
	 *
	 * @type {string}
	 * @memberof valueComponentProperty
	 */
	selectedText: string;

	/**
	 * 是否展開
	 *
	 * @type {boolean}
	 * @memberof valueComponentProperty
	 */
	isOpen: boolean;

}

const ValueComponent: React.FC<valueComponentProperty> = ({ selectedText, isOpen }) => (
	<div className={styles.dropdownSelectField}>
		<input
			type="text"
			disabled
			value={selectedText || ''}
		/>
		<div className={classnames(styles.icon, {
			[styles.rotate]: isOpen
		})}>
			<ExpandDown />
		</div>
	</div>
);

interface panelComponentProperty {

	/**
	 * 選項列表
	 *
	 * @type {OptionBase[]}
	 * @memberof panelComponentProperty
	 */
	options: OptionBase[],

	/**
	 * 選項變更時觸發的函式
	 *
	 * @memberof panelComponentProperty
	 */
	onChangeValue: (val: string) => void

}

const PanelComponent: React.FC<panelComponentProperty> = ({
	options = [],
	onChangeValue = () => {}
}) => (
	<ul className={styles.dropdownPanel}>
		{
			options.map(({ id, name }) => (
				<li key={id}>
					<div
						role="button"
						tabIndex={0}
						onKeyPress={() => {}}
						key={id}
						onClick={() => onChangeValue(id)}
					>
						{name}
					</div>
				</li>
			))
		}
	</ul>
);

interface DropdownSelectProperty {

	/**
	 * 下拉選項列表
	 *
	 * @type {OptionBase[]}
	 * @memberof DropdownSelectProperty
	 */
	options: OptionBase[],

	/**
	 * 選中的選項
	 *
	 * @type {OptionBase}
	 * @memberof DropdownSelectProperty
	 */
	selectedId: string,

	/**
	 * 變更選項時觸法的函式
	 *
	 * @memberof DropdownSelectProperty
	 */
	onChangeValue?: (val: string) => void,

}

const DropdownSelect: React.FC<DropdownSelectProperty> = ({
	options,
	selectedId,
	onChangeValue = () => {}
}) => {

	const selectOption = options.find(option => option.id === selectedId);
	const selectedText = selectOption ? selectOption.name : '';

	return (
		<Dropdown
			valueComponent={({ isOpen }) => (
				<ValueComponent
					selectedText={selectedText}
					isOpen={isOpen}
				/>
			)}
			panelComponent={() => (
				<PanelComponent
					options={options}
					onChangeValue={onChangeValue}
				/>
			)}
		/>
	);
};

export default DropdownSelect;
