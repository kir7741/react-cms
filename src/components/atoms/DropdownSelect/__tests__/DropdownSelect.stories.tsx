import React from 'react';
import { storiesOf } from '@storybook/react';

import DropdownSelect from 'components/atoms/DropdownSelect';

const stories = storiesOf('atoms/DropdownSelect', module);

stories.add('__interactive', () => <DropdownSelect options={[]} selectedId=''/>);
