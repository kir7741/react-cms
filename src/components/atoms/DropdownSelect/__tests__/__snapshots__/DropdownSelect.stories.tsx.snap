// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`__interactive 1`] = `
<DocumentFragment>
  <div
    class="dropdownWrapper"
  >
    <div
      role="button"
      tabindex="0"
    >
      <div
        class="dropdownSelectField"
      >
        <input
          disabled=""
          type="text"
          value=""
        />
        <div
          class="icon"
        >
          <test-file-stub />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
